torch/distributed/traceable_collectives.py

```diff
@@ -0,0 +1,207 @@
+import weakref
+
+import torch
+import torch.distributed as dist
+
+from torch._C import _disabled_torch_function_impl
+from torch.utils._pytree import tree_map
+from typing import Any, Tuple, Union, List, cast
+
+import torch.distributed.distributed_c10d as c10d
+import torch.distributed._tensor as dt
+
+"""
+New traceable, functional collectives.
```

Comment:
I wonder if it makes more sense to rename this file to `functional_collectives.py`. Since we expect users to import this as the new standard, this naming might make more sense to help users interpret these as functions that "do not modify data" and are the successor to our old collectives which do modify tensors inplace. Traceability is more of a design decision on our part and an average user wont have much context there?

Reply:
Makes sense, I'll move it under `_functional_collectives.py` to reflect the experimental nature of them.